org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -373,5 +405,29 @@ protected void doClean(ToBeBuilt toBeBuilt, IProgressMonitor monitor) throws Cor
 			progress.worked(1);
 		}
 	}
-	
+
+	@Override
+	public ISchedulingRule getRule(int kind, Map<String, String> args) {
+		switch (getSchedulingOption()) {
+			case NULL: return null;
+			case WORKSPACE: return getProject().getWorkspace().getRoot();
+			case PROJECT: return getProject();
+			case ALL_XTEXT_PROJECTS: return new MultiRule(Arrays.stream(
+				getProject().getWorkspace().getRoot().getProjects())
+				.filter(XtextProjectHelper::hasNature)
+				.toArray(ISchedulingRule[]::new));
+			default: throw new IllegalArgumentException();
+		}
+	}
+
+	/**
+	 * @since 2.16
+	 */
+	private SchedulingOption getSchedulingOption () {
+		String schedulingRuleName = preferenceStore.getString(PREF_SCHEDULING_RULE);
```

Comment:
Does this always do i/o or is this using cached information?

Reply:
I assume this is uncached, would have to test that. Do you think it would be worth to store the value and listen to changes? Would it make sense to enable IPreferenceStoreAccess to be used for such scenarios? (Currently it can only be used in project context)

Reply:
When we would like to enable change by a pref page, we have to react on changes.

Reply:
Yes we have preference change listeners all over the place

Reply:
Ok, then I‘ll add a field that updates on change